### PR TITLE
Align Report with other Oss Index tools

### LINKED ISSUE #43: [FEATURE] Change default output for a table structure
* **What are you trying to do?**

The current output is a tree/graph showing the dependencies (direct and transitive) and below the vulnerabilities found. An example can be found at: https://github.com/sonatype-nexus-community/scan-gradle-plugin/pull/34

To keep consistency with other OSS Index tools, and make the output easier to read there is the initiative to make the tree/graph as an optional output (requiring a new flag) and use as default a table structure.

Screen Shot 2020-08-20 at 6 08 39 PM

At the end, if any vulnerability is found show a hint on how to get the tree output to start fixing vulnerabilities.

cc @bhamail / @DarthHater / @guillermo-varela

### PR DESCRIPTION
**Pull request by shaikhu**
1. Align report with other Oss Index tools (specifically Audit JS)
2. Make this report the default.
3. Allow dependencyGraph report to be available through flag.

It relates to the following issue #s:
* Fixes #43 

cc @guillermo-varela

image

image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -67,7 +67,8 @@ ossIndexAudit {
     useCache = true // true by default
     cacheDirectory = 'some/path' // by default it uses the user data directory (according to OS)
     cacheExpiration = 'PT12H' // 12 hours if omitted. It must follow the Joda Time specification at https://www.javadoc.io/doc/joda-time/joda-time/2.10.4/org/joda/time/Duration.html#parse-java.lang.String-
-    colorEnabled = false // if true prints vulnerability description in color. By default is true.  
+    colorEnabled = false // if true prints vulnerability description in color. By default is true.
+    dependencyGraph = true// if true prints dependency graph showing direct/transitive dependencies. By defaul is false.
```
Comment (guillermo-varela) [inline README.md:71]: Minor: `true//` -> `true //`

File: src/integTest/java/org/sonatype/gradle/plugins/scan/ScanPluginIntegrationTestBase.java [hunk 2]
```diff
@@ -221,7 +253,7 @@ public void testAuditTask_IncludeTestDependencies_OssIndex() throws IOException
         .withArguments("ossIndexAudit", "--info")
```
Comment (guillermo-varela) [inline src/integTest/java/org/sonatype/gradle/plugins/scan/ScanPluginIntegrationTestBase.java:253]: Both `testAuditTask_ExcludeTestDependencies_OssIndex` and `testAuditTask_IncludeTestDependencies_OssIndex` should have tests using the graph and the table to make sure we're not missing dependencies in such cases.

File: src/main/java/org/sonatype/gradle/plugins/scan/ossindex/BannerUtils.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      log.error(e.getMessage(), e);
+    }
+    return sb.toString();
+  }
+
+  private static String getVersion() throws IOException {
+    Properties properties = new Properties();
+    properties.load(BannerUtils.class.getClassLoader().getResourceAsStream(PROPERTIES_PATH));
+    return properties.getProperty("version");
+  }
```
Comment (guillermo-varela) [inline src/main/java/org/sonatype/gradle/plugins/scan/ossindex/BannerUtils.java:60]: Would it be possible to have another class where we can get the version from?
We're already reading that property file elsewhere to extract the version so it'd be helpful to have a single place to get it: https://github.com/sonatype-nexus-community/scan-gradle-plugin/blob/52a8e43f474273310a32a91980c9bafa2c5b6a2f/src/main/java/org/sonatype/gradle/plugins/scan/ossindex/TransportBuilder.java#L47-L50

File: src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DefaultResponseHandler.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+      packageUrlProcessingText += "No vulnerabilities found!";
+    }
+    else if (vulnerabilities.size() == 1) {
+      packageUrlProcessingText += "1 vulnerability found!";
+    }
+    else {
+      packageUrlProcessingText += vulnerabilities.size() + " vulnerabilities found!";
+    }
+
+
```
Comment (guillermo-varela) [inline src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DefaultResponseHandler.java:101]: Minor: remove extra line

File: src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DefaultResponseHandler.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+  private String addColour(Float cvssScore, String text) {
+    if (extension.isColorEnabled()) {
+      return VulnerabilityUtils.addColorBasedOnCvssScore(cvssScore, text);
+    }
+    else {
+      return text;
+    }
+  }
+}
```
Comment (guillermo-varela) [inline src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DefaultResponseHandler.java:154]: Minor: new line at the end

File: src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DependencyGraphResponseHandler.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+    StringBuilder vulnerabilitiesText = new StringBuilder()
+        .append(vulnerabilities.size())
+        .append(" vulnerabilities detected");
+
+    vulnerabilitiesText.append(vulnerabilities.stream()
+        .map(vulnerability -> handleComponentReportVulnerability(vulnerability, prefix))
+        .collect(Collectors.joining("")));
+
+
```
Comment (guillermo-varela) [inline src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DependencyGraphResponseHandler.java:96]: Minor: remove extra line

File: src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DependencyGraphResponseHandler.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    String indent = System.lineSeparator() +
+        StringUtils.replaceOnce(prefix, DEPENDENCY_PREFIX, StringUtils.repeat(" ", DEPENDENCY_PREFIX.length()));
+
+    StringBuilder builder = new StringBuilder()
+        .append(vulnerability.getTitle());
+
+    Float cvssScore = vulnerability.getCvssScore();
+    if (cvssScore != null) {
+      builder.append(" (").append(cvssScore).append("/10").append(", ")
+          .append(VulnerabilityUtils.getAssessment(cvssScore)).append(')');
```
Comment (guillermo-varela) [inline src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DependencyGraphResponseHandler.java:141]:
```suggestion
      builder.append(" (")
          .append(cvssScore)
          .append("/10")
          .append(", ")
          .append(VulnerabilityUtils.getAssessment(cvssScore))
          .append(')');
```

File: src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DependencyGraphResponseHandler.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+    builder.append(": ").append(vulnerability.getReference());
+
+    String vulnerabilityText = builder.toString();
+    if (extension.isColorEnabled()) {
+      vulnerabilityText = VulnerabilityUtils.addColorBasedOnCvssScore(cvssScore, vulnerabilityText);
+    }
+    return indent + vulnerabilityText;
+  }
+}
```
Comment (guillermo-varela) [inline src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DependencyGraphResponseHandler.java:152]: Minor: new line at the end of file

File: src/main/java/org/sonatype/gradle/plugins/scan/ossindex/OssIndexAuditTask.java [hunk 9]
```diff
@@ -99,7 +83,14 @@ public void audit() {
...
 
-      hasVulnerabilities = handleOssIndexResponse(dependencies, dependenciesMap, response);
+      if (extension.isDependencyGraph()) {
+        hasVulnerabilities = new DependencyGraphResponseHandler(extension)
+            .handleOssIndexResponse(dependencies, dependenciesMap, response);
+      }
+      else {
+        hasVulnerabilities = new DefaultResponseHandler(extension)
+            .handleOssIndexResponse(dependencies, dependenciesMap, response);
+      }
```
Comment (guillermo-varela) [inline src/main/java/org/sonatype/gradle/plugins/scan/ossindex/OssIndexAuditTask.java:93]:
```suggestion
      OssIndexResponseHandler responseHandler = extension.isDependencyGraph()
              ? new DependencyGraphResponseHandler(extension)
              : new DefaultResponseHandler(extension);
      hasVulnerabilities = responseHandler.handleOssIndexResponse(dependencies, dependenciesMap, response);
```

File: src/main/java/org/sonatype/gradle/plugins/scan/ossindex/VulnerabilityUtils.java [hunk 10]
```diff
@@ -17,17 +17,19 @@
 
 import java.util.Objects;
 
-public class VulnerabilityUtils
+class VulnerabilityUtils
```
Comment (guillermo-varela) [inline src/main/java/org/sonatype/gradle/plugins/scan/ossindex/VulnerabilityUtils.java:20]: Why did the access modifier change?

File: src/main/resources/org/sonatype/gradle/plugins/scan/ossindex/header.txt [hunk 11]
```diff
@@ -0,0 +1,15 @@
+ ________  ________  ________  ________  ___       _______           ________  ________  ________  ________
```
Comment (guillermo-varela) [inline src/main/resources/org/sonatype/gradle/plugins/scan/ossindex/header.txt:1]: I'd prefer this file to be called "banner.txt" instead for two reasons:
1. The class printing this is `BannerUtils`
2. The file we use internally at Sonatype for the license header is `header.txt` (to use a Maven plugin I believe), so this file might create confusions.

File: src/test/java/org/sonatype/gradle/plugins/scan/ossindex/OssIndexAuditTaskTest.java [hunk 12]
```diff
@@ -99,6 +99,7 @@ private OssIndexAuditTask buildAuditTaskSpy(boolean isSimulated) {
 
     OssIndexPluginExtension extension = new OssIndexPluginExtension(project);
     extension.setSimulationEnabled(isSimulated);
+    extension.setDependencyGraph(true);
```
Comment (guillermo-varela) [inline src/test/java/org/sonatype/gradle/plugins/scan/ossindex/OssIndexAuditTaskTest.java:102]: Why was this necessary?
Is the message `Vulnerabilities detected, check log output to review them` gone in the new output?

File: README.md (hunk 1 above)
Comment (shaikhu) [reply README.md]: > Minor: `true//` -> `true //`

Done. 41a8f3bf

File: src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DefaultResponseHandler.java (hunk 5 above)
Comment (shaikhu) [reply src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DefaultResponseHandler.java]: > Minor: new line at the end

Done. f1356f24

File: src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DependencyGraphResponseHandler.java (hunk 6 above)
Comment (shaikhu) [reply src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DependencyGraphResponseHandler.java]: > Minor: remove extra line

Done. fd11eea2

File: src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DependencyGraphResponseHandler.java (hunk 8 above)
Comment (shaikhu) [reply src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DependencyGraphResponseHandler.java]: > Minor: new line at the end of file

Done. 8b051ea4

File: src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DefaultResponseHandler.java (hunk 4 above)
Comment (shaikhu) [reply src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DefaultResponseHandler.java]: > Minor: remove extra line

Done. 5f11b065

File: src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DependencyGraphResponseHandler.java (hunk 7 above)
Comment (shaikhu) [reply src/main/java/org/sonatype/gradle/plugins/scan/ossindex/DependencyGraphResponseHandler.java]: > 

Github won't let me commit suggestion so committed manually. b719e451

File: src/main/resources/org/sonatype/gradle/plugins/scan/ossindex/header.txt (hunk 11 above)
Comment (shaikhu) [reply src/main/resources/org/sonatype/gradle/plugins/scan/ossindex/header.txt]: > I'd prefer this file to be called "banner.txt" instead for two reasons:
> 
> 1. The class printing this is `BannerUtils`
> 2. The file we use internally at Sonatype for the license header is `header.txt` (to use a Maven plugin I believe), so this file might create confusions.

Done. 36701e65

File: src/main/java/org/sonatype/gradle/plugins/scan/ossindex/VulnerabilityUtils.java (hunk 10 above)
Comment (shaikhu) [reply src/main/java/org/sonatype/gradle/plugins/scan/ossindex/VulnerabilityUtils.java:20]: > Why did the access modifier change?

1. The class isn't accessed outside of the `ossindex` package.
2. Make consistent with the other utils (`BannerUtils`) class.

File: src/main/java/org/sonatype/gradle/plugins/scan/ossindex/BannerUtils.java (hunk 3 above)
Comment (shaikhu) [reply src/main/java/org/sonatype/gradle/plugins/scan/ossindex/BannerUtils.java]: > Would it be possible to have another class where we can get the version from?
> We're already reading that property file elsewhere to extract the version so it'd be helpful to have a single place to get it:
> 
> https://github.com/sonatype-nexus-community/scan-gradle-plugin/blob/52a8e43f474273310a32a91980c9bafa2c5b6a2f/src/main/java/org/sonatype/gradle/plugins/scan/ossindex/TransportBuilder.java#L47-L50

Done. 3f2bd797

File: src/integTest/java/org/sonatype/gradle/plugins/scan/ScanPluginIntegrationTestBase.java (hunk 2 above)
Comment (shaikhu) [reply src/integTest/java/org/sonatype/gradle/plugins/scan/ScanPluginIntegrationTestBase.java:268]: > Both `testAuditTask_ExcludeTestDependencies_OssIndex` and `testAuditTask_IncludeTestDependencies_OssIndex` should have tests using the graph and the table to make sure we're not missing dependencies in such cases.

Done bac82216

File: src/test/java/org/sonatype/gradle/plugins/scan/ossindex/OssIndexAuditTaskTest.java (hunk 12 above)
Comment (shaikhu) [reply src/test/java/org/sonatype/gradle/plugins/scan/ossindex/OssIndexAuditTaskTest.java]: > Is the message `Vulnerabilities detected, check log output to review them` gone in the new output?

No, we still included that message.

> Why was this necessary?

I should of looked at the error message closer. For the default/table report, a `NullPointerException` was being thrown because we don't set the vulnerability description in the test setup (description isn't used in the dependency graph report). I've add a default value in case of null (shouldn't happen though ;-)). See b1836685

**Review by guillermo-varela — APPROVED** [review]
+1
